hollow/src/main/java/com/netflix/hollow/api/consumer/index/PrimaryKeyIndex.java

```diff
@@ -0,0 +1,319 @@
+package com.netflix.hollow.api.consumer.index;
+
+import static java.util.stream.Collectors.toList;
+
+import com.netflix.hollow.api.consumer.HollowConsumer;
+import com.netflix.hollow.api.custom.HollowAPI;
+import com.netflix.hollow.api.objects.HollowObject;
+import com.netflix.hollow.core.index.FieldPaths;
+import com.netflix.hollow.core.index.HollowPrimaryKeyIndex;
+import com.netflix.hollow.core.index.key.PrimaryKey;
+import com.netflix.hollow.core.read.engine.HollowReadStateEngine;
+import com.netflix.hollow.core.schema.HollowObjectSchema;
+import com.netflix.hollow.core.schema.HollowSchema;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+/**
+ * A type safe index for indexing with a primary, or unique, key.
+ *
+ * @param <T> the primary type
+ * @param <Q> the key type
+ */
+public class PrimaryKeyIndex<T extends HollowObject, Q> {
```

Comment:
Should this be `PrimaryKeyIndex` or `UniqueKeyIndex`?

@davesu believes that `PrimaryKeyIndex` is a misleading name. It implies a strong connection to the `@HollowPrimaryKey` annotation, of which there can be one per type. Whereas a type could have more than one unique key index (and also not have a primary key defined at all).

Options I can think of:

* **as-is in this PR**. Go with this PR as written. Address the nuance of primary versus unique keys through documentation
* **Rename this class to `UniqueKeyIndex`**. Keep constructors or helpers when you have a `PrimaryKey`
* **Introduce `UniqueKeyIndex` superclass**. This `PrimaryKeyIndex` would subclass and exist just for the `PrimaryKey` constructors or helpers.